src/python/dxpy/bindings/download_all_inputs.py

```diff
@@ -0,0 +1,168 @@
+
+def _get_downloaded_files(inputs, idir):
+    downloaded_files = defaultdict(list)
+
+    for k in inputs:
+        for f in inputs[k]:
+            downloaded_files[k].append(os.path.join(idir, f['trg_fname']))
+
+    return downloaded_files
+
+def download_all_inputs(exclude=None, parallel=True, maxThreads=None):
+    '''
+    :param exclude: List of input variables that should not be downloaded.
+    :type exclude: Array of strings
+    :param parallel: Should we download multiple files in parallel? (default: True)
+    :type filename: boolean
+    :param maxThreads: If parallel is True, how many threads should be used to download files? (default: number of cores)
+    :type append: int
+
+    returns a : dict of lists of strings where each key is the input variable
+                and each list element is the full path to the file that has
+                been downloaded.
+
+    This function downloads all files that were supplied as inputs to the app.
+    By convention, if an input parameter "FOO" has value
+
+        {"$dnanexus_link": "file-xxxx"}
+
+    and filename INPUT.TXT, then the linked file will be downloaded into the
+    path:
+
+        $HOME/in/FOO/INPUT.TXT
+
+    If an input is an array of files, then all files will be placed into
+    numbered subdirectories under a parent directory named for the
+    input. For example, if the input key is FOO, and the inputs are {A, B,
+    C}.vcf then, the directory structure will be:
+
+        $HOME/in/FOO/0/A.vcf
+                     1/B.vcf
+                     2/C.vcf
+
+    Zero padding is used to ensure argument order. For example, if there are
+    12 input files {A, B, C, D, E, F, G, H, I, J, K, L}.txt, the directory
+    structure will be:
+
+        $HOME/in/FOO/00/A.vcf
+                     ...
+                     11/L.vcf
+
+    This allows using shell globbing (FOO/*/*.vcf) to get all the files in the input
+    order and prevents issues with files which have the same filename.'''
+
+    # Input directory, where all inputs are downloaded
+    idir = file_load_utils.get_input_dir()
+    try:
+        job_input_file = file_load_utils.get_input_json_file()
+        dirs, inputs, rest = file_load_utils.get_job_input_filenames(job_input_file)
+    except IOError:
+        msg = 'Error: Could not find the input json file: {0}.\n'.format(job_input_file)
+        msg += '       This function should only be called from within a running job.'
+        print(msg)
+        raise
+
+    # Exclude directories
+    dirs_to_create = []
+    for d in dirs:
+        if (exclude is None) or (d not in exclude):
+            dirs_to_create.append(d)
+
+    # Create the directory structure, in preparation for download.
+    # Allows performing the download in parallel.
+    _create_dirs(idir, dirs_to_create)
+
+    # Remove excluded inputs
+    if (exclude is not None) and len(exclude) > 0:
+        inputs = file_load_utils.filter_dict(inputs, args.exclude)
```

Comment:
I believe `args.exclude` is an invalid reference here. To catch errors like this, you can run `pylint -E <file.py>` and/or `pyflakes <file.py>`.